### ISSUE TITLE
gu proxy issue

### ISSUE THREAD

**Issue by wlnirvana** [oracle/graal#3175]
**Describe GraalVM and your environment :**
 - GraalVM version or commit id if built from source: **20.3.0**
 - CE or EE: **CE**
 - JDK version: **JDK11**
 - OS and OS Version: **Ubuntu 20.04**
 - Architecture: **amd64**
 - The output of `java -Xinternalversion`: 
```
OpenJDK 64-Bit Server VM (11.0.9+10-jvmci-20.3-b06) for linux-amd64 JRE (11.0.9+10-jvmci-20.3-b06), built on Nov 12 2020 06:22:27 by "buildslave" with gcc 7.3.0
```

Also the output of `cat $GRAALVM_HOME/release`:

```
JAVA_VERSION="11.0.9"
MODULES="java.base java.logging jdk.internal.vm.ci jdk.unsupported org.graalvm.sdk java.datatransfer java.xml java.prefs java.desktop java.management org.graalvm.truffle jdk.internal.vm.compiler com.oracle.graal.graal_enterprise java.compiler java.instrument java.security.sasl java.naming java.rmi java.management.rmi java.net.http java.scripting java.security.jgss java.transaction.xa java.sql java.sql.rowset java.xml.crypto java.se java.smartcardio jdk.accessibility jdk.aot jdk.internal.jvmstat jdk.attach jdk.charsets jdk.compiler jdk.crypto.ec jdk.crypto.cryptoki jdk.dynalink jdk.internal.ed jdk.editpad jdk.hotspot.agent jdk.httpserver jdk.internal.le jdk.internal.opt jdk.management jdk.internal.vm.compiler.management jdk.jfr jdk.internal.vm.compiler.truffle.jfr jdk.jartool jdk.javadoc jdk.jcmd jdk.management.agent jdk.jconsole jdk.jdeps jdk.jdwp.agent jdk.jdi jdk.jlink jdk.jshell jdk.jsobject jdk.jstatd jdk.localedata jdk.management.jfr jdk.naming.dns jdk.naming.ldap jdk.naming.rmi jdk.net jdk.pack jdk.rmic jdk.scripting.nashorn jdk.scripting.nashorn.shell jdk.sctp jdk.security.auth jdk.security.jgss jdk.unsupported.desktop jdk.xml.dom jdk.zipfs org.graalvm.js.scriptengine org.graalvm.locator"
OS_NAME="linux"
OS_ARCH="amd64"
GRAALVM_VERSION="20.3.0"
SOURCE="compiler:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 graal-js:b53c4e144b5909876b377e31be69e3b3f381ae0b graal-nodejs:b53c4e144b5909876b377e31be69e3b3f381ae0b java-benchmarks:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 regex:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 sdk:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 substratevm:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 sulong:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 tools:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 truffle:c5ff5a5a091bfd974640aaf6fbe51c81bd080438 vm:c5ff5a5a091bfd974640aaf6fbe51c81bd080438"
COMMIT_INFO={"compiler": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "graal-js": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217687, "commit.rev": "b53c4e144b5909876b377e31be69e3b3f381ae0b"}, "graal-nodejs": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217687, "commit.rev": "b53c4e144b5909876b377e31be69e3b3f381ae0b"}, "java-benchmarks": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "regex": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "sdk": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "substratevm": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "sulong": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "tools": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "truffle": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}, "vm": {"commit.committer": "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts": 1605217754, "commit.rev": "c5ff5a5a091bfd974640aaf6fbe51c81bd080438"}}
component_catalog="https://www.graalvm.org/component-catalog/graal-updater-component-catalog-java11.properties"
```

**Have you verified this issue still happens when using the latest snapshot?**
Yes. With [the latest version][1], I got the same error message.

[1]:https://github.com/graalvm/graalvm-ce-dev-builds/releases/download/21.1.0-dev-20210126_1929/graalvm-ce-java11-linux-amd64-dev.tar.gz

**Describe the issue**
Using `gu` behind a proxy does not work.

**Code snippet or code repository that reproduces the issue**
```
$ env | grep proxy
http_proxy=127.0.0.1:8001
https_proxy=127.0.0.1:8001

$ curl https://google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="https://www.google.com/">here</A>.

$ ./gu available
Downloading: Component catalog from www.graalvm.org
Error: Component catalog is unreachable. 
Please check your connection and proxy settings. If your machine is behind a proxy, environment variables (http_proxy, https_proxy, ...) must be set appropriately.
```

However, following [the instruction here][2] to manually specify the jvm options works:

[2]:https://github.com/oracle/graal/issues/617#issuecomment-733852267

```
$ ./gu --vm.Dhttp.proxyHost=127.0.0.1 --vm.Dhttp.proxyPort=8001 --vm.Dhttps.proxyHost=127.0.0.1 --vm.Dhttps.proxyPort=8001 available
Downloading: Component catalog from www.graalvm.org
ComponentId              Version             Component name      Origin 
--------------------------------------------------------------------------------
llvm-toolchain           20.3.0              LLVM.org toolchain  github.com
native-image             20.3.0              Native Image        github.com
python                   20.3.0              Graal.Python        github.com
R                        20.3.0              FastR               github.com
ruby                     20.3.0              TruffleRuby         github.com
wasm                     20.3.0              GraalWasm           github.com
```

**Comment by maassg** [comment]
for me on windows 10 it does not work with proposed proxy settings:

I did a successful short test with a java program to reach to the url:
https://www.graalvm.org/component-catalog/v2/graal-updater-component-catalog-java11.properties
and it worked with the following properties set:
- Djava.net.useSystemProxies=true
- Djavax.net.ssl.trustStore=C:/Windows/Sun/Java/Deployment/cacerts

But trying to call gu.cmd install native-image with the same properties set
does still not work:

gu.cmd --jvm
--vm.Djava.net.useSystemProxies=true
--vm.Djavax.net.ssl.trustStore=C:/Windows/Sun/Java/Deployment/cacerts
install native-image

Error: Error downloading component catalog from https://www.graalvm.org/component-catalog/v2/graal-updater-component-catalog-java11.properties: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target.
Please check your connection and proxy settings. If your machine is behind a proxy, environment variables (http_proxy, https_proxy, ...) must be set appropriately.

**Comment by mcraj017** [comment]
@wlnirvana thanks for reporting the issue, I tried the same on my pc, and it worked
` $./gu available
Downloading: Component catalog from www.graalvm.org
Error: Component catalog is unreachable. 
Please check your connection and proxy settings. If your machine is behind a proxy, environment variables (http_proxy, https_proxy, ...) must be set appropriately.
$ export https_proxy=<proxy>
$ ./gu available
Downloading: Component catalog from www.graalvm.org
ComponentId              Version             Component name      Origin 
--------------------------------------------------------------------------------
llvm-toolchain           20.3.0              LLVM.org toolchain  github.com
native-image             20.3.0              Native Image        github.com
python                   20.3.0              Graal.Python        github.com
R                        20.3.0              FastR               github.com
ruby                     20.3.0              TruffleRuby         github.com
wasm                     20.3.0              GraalWasm           github.com`

**Comment by belu** [comment]
This could be related to https://github.com/oracle/graal/issues/3173.

A workaround is mentioned there: run `gu` in JVM mode: `gu --jvm ...`

**Comment by sdedic** [comment]
@wlnirvana please try to confirm the case is different from #3173 and  #3175; run the command as `gu -e` to see the underlying connection/transport error. Please post the error stacktrace here, if possible. 
If the error turns out to be thrown from cryptography lib,  try to run with `--jvm` (but do not specify the `useSystemProxies` for now) - if the cause is same as in the other issues, it will disappear.

**Comment by wlnirvana** [comment]
@sdedic 

Due to network changes, I can access the catalog now directly without using a proxy. However, I doubt this is the same issue with #3173 because

1. the error reported is "Component catalog is unreachable." instead of no such algorithm
2. By monitoring the proxy log, it can be seen that `./gu --vm.Dhttp.proxyHost=127.0.0.1 --vm.Dhttp.proxyPort=8001 --vm.Dhttps.proxyHost=127.0.0.1 --vm.Dhttps.proxyPort=8001 available` results in a log being printed `2021/02/20 08:44:59 127.0.0.1:51192 accepted //www.graalvm.org:443`
3. `https_proxy=127.0.0.1:8001 ./gu --vm.Djava.net.useSystemProxy=true available` won't print any log in proxy software.
4. `https_proxy=127.0.0.1:8001 ./gu --jvm --vm.Djava.net.useSystemProxy=true available` won't either.

**Comment by maassg** [comment]
Version: GraalVM Community 21.0.0
OS: Windows 10 behind proxy

1. gu.cmd
gu with environment variables "http_proxy and https_proxy" does not work at all
the variables are definitely not recognized
after timeout the output says:
... Component catalog is unreachable.
Please check your connection and proxy settings. If your machine is behind a proxy, environment variables (http_proxy, https_proxy, ...) must be set appropriately.

2. gu.cmd with --jvm  and:
--vm.Dhttp.proxyHost=<...>
--vm.Dhttp.proxyPort=<...>
--vm.Dhttps.proxyHost=<...>
--vm.Dhttps.proxyPort=<...>
install native-image
results in a immediate response:
 
Downloading: Component catalog from www.graalvm.org
javax.net.ssl|DEBUG|0F|pool-1-thread-1|2021-02-20 10:49:55.445 CET|SSLCipher.java:438|jdk.tls.keyLimits:  entry = AES/GCM/NoPadding KeyUpdate 2^37. AES/GCM/NOPADDING:KEYUPDATE = 137438953472
javax.net.ssl|ERROR|0F|pool-1-thread-1|2021-02-20 10:49:55.823 CET|TransportContext.java:341|Fatal (CERTIFICATE_UNKNOWN): PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

3. same as 2. but additionally with --vm.Djavax.net.ssl.trustStore
gives the same result

4. downloading https://www.graalvm.org/component-catalog/v2/graal-updater-component-catalog-java11.properties with a simple POJO and --vm.Djavax.net.ssl.trustStore set
works as a charm
without --vm.Djavax.net.ssl.trustStore set 
results in the same error as under 2. and 3.:

URL: https://www.graalvm.org/component-catalog/v2/graal-updater-component-catalog-java11.properties
javax.net.ssl|DEBUG|01|main|2021-02-20 11:03:43.931 CET|SSLCipher.java:465|jdk.tls.keyLimits:  entry = AES/GCM/NoPadding KeyUpdate 2^37. AES/GCM/NOPADDING:KEYUPDATE = 137438953472
>>> URLConnection: sun.net.www.protocol.https.DelegateHttpsURLConnection:https://www.graalvm.org/component-catalog/v2/graal-updater-component-catalog-java11.properties
User-Agent=[Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; AS; rv:11.0) like Gecko]
javax.net.ssl|ERROR|01|main|2021-02-20 11:03:44.402 CET|TransportContext.java:361|Fatal (CERTIFICATE_UNKNOWN): PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

CONCLUSION:
it looks like the gu-tool does not consider / interpret the indispensable trustStore setting!

**Comment by sdedic** [comment]
> 3. `https_proxy=127.0.0.1:8001 ./gu --vm.Djava.net.useSystemProxy=true available` won't print any log in proxy software.
> 4. `https_proxy=127.0.0.1:8001 ./gu --jvm --vm.Djava.net.useSystemProxy=true available` won't either.

OK, 
- first try to use full URL, so `https_proxy=http://127.0.0.1:8001 ./gu ....`. The tool should accept both syntaxes (it does on my Linux), but just to be sure
- I repeat: **DO NOT** use `--vm.Djava.net.useSystemProxy=true`. This will setup JVM network stack to use OS-native integration, and who knows how broken it can be on Windows OS or how it mingles with GU's implementation. Let's diagnose the problem piece by piece, so 
- please run something like `https_proxy=http://127.0.0.1:8001 ./gu available`, so no system/JVM level stuff is involved. Please observe your proxy's log - that's a perfect tool, thanks for the idea !

**Comment by sdedic** [comment]
@maassg 

> 1. gu.cmd
>    gu with environment variables "http_proxy and https_proxy" does not work at all
>    the variables are definitely not recognized

This is hard to believe, as there are tests that specifically simulate the proxies are contacted (well, that at least `URLStreamHandler::openConnection(URL u, Proxy p)` is called)... but will test after I get to a Windows installation. Could you please share:
- the exact OS version,
- the exact GraalVM version,
- the env vars effective in our commandline (`set http` should print out all stuff that begin with "http")
- ideally the exact commandline (please mangle just your internal network URLs)
- 
I wonder if the `--jvm` can have a windows-specific bug...

**Comment by wlnirvana** [comment]
@sdedic 

`https_proxy=http://127.0.0.1:8001 ./gu available` works and prints the proxy log. Thanks for your advice!

`https_proxy= 127.0.0.1:8001 ./gu available` does not use the proxy, as previously reported.

Should this issue be closed? Or this is unexpected and needs to be fixed?

**Comment by sdedic** [comment]
OK, after some time experimenting ... yes, there are some issues. You can verify, if you'd like.

First, `java.net.useSystemProxies=true` appears to be broken for binaries compiled by native-image (which includes `gu`). In my setup: reported to GraalVM / native-image team. In JVM mode (`gu --jvm`), the tool reads system settings at least for manually specified proxy. I was not able to verify it works with automatic proxy detection (it does not work in my current VirtualBox Windows setup at all :))

Second: yes, there are bugs in GU ! Subtle and funny, but here it is: you pasted the executed commandline:
> https_proxy= 127.0.0.1:8001 ./gu available

Note the space between `=` and the proxy address ... and that's it. GU should trim the string it reads from the env variable. In addition, the code that parses the variable fails on **numeric host with no protocol**. So while `localhost:8001` would work, as well as `http://127.0.0.1:8001`, `127.0.0.1:8001` fails. I filed these for myself to fix.

Further, the function of `--vm.Djava.net.useSystemProxies=true` depends on OS and JDK version: on Windows, JDK 8 **only** reads manual proxy configuraiton. For the automatic (PAC) configuraiton, JDK 9+ is needed :)

Please leave the issue open, I'll use it to track the GU fixes.

**Comment by sdedic** [comment]
Sadly I still wasn't able to reproduce the -Dhttp.proxyHost issue reported above.

**Comment by wlnirvana** [comment]
Oh, I did not even notice that I had a space there. It should be a typo that I accidentally keyed in after pasting. Anyway, I just double checked `https_proxy=127.0.0.1:8001 ./gu available`. Still not working, no logs printed in proxy app. But just as you said, `localhost:8001` works. So now there are two valid options:

1. `https_proxy=localhost:8001 ./gu available`
2. `https_proxy=http://127.0.0.1:8001 ./gu available`

**Comment by sdedic** [comment]
Should be now fixed by 33bfa104
